mlir/lib/Quantum/Transforms/VerifyParentGateAnalysis.hpp

```diff
@@ -0,0 +1,171 @@
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+
+template <typename OpType> class VerifyParentGateAnalysis {
+  public:
+    VerifyParentGateAnalysis(OpType gate)
+    {
+        ValueRange inQubits = gate.getInQubits();
+        auto parentGate = dyn_cast_or_null<OpType>(inQubits[0].getDefiningOp());
+
+        if (!verifyParentGateType(gate, parentGate)) {
+            setVerifierResult(false);
+            return;
+        }
+
+        if (!verifyAllInQubits(gate, parentGate)) {
+            setVerifierResult(false);
+            return;
+        }
+    }
+
+    bool getVerifierResult() { return succeeded; }
+
+    void setVerifierResult(bool b) { succeeded = b; }
+
+  private:
+    bool succeeded = true;
+
+    bool verifyParentGateType(OpType op, OpType parentOp) const
+    {
+        // Verify that the parent gate is of the same type.
+        // If OpType is quantum.custom, also verify that parent gate has the
+        // same gate name.
+
+        if (!parentOp || !isa<OpType>(parentOp)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    bool verifyAllInQubits(OpType op, OpType parentOp) const
+    {
+        // Verify that parent's results and current gate's inputs are in the same order
+        // If the gates are controlled, both gates' control wires and values
+        // must be the same. The control wires must be in the same order.
+
+        ValueRange inNonCtrlQubits = op.getNonCtrlQubitOperands();
+        ValueRange inCtrlQubits = op.getCtrlQubitOperands();
+        ValueRange parentOutNonCtrlQubits = parentOp.getNonCtrlQubitResults();
+        ValueRange parentOutCtrlQubits = parentOp.getCtrlQubitResults();
+
+        if ((inNonCtrlQubits.size() != parentOutNonCtrlQubits.size()) ||
+            (inCtrlQubits.size() != parentOutCtrlQubits.size())) {
+            return false;
+        }
+
+        for (const auto &[idx, qubit] : llvm::enumerate(inNonCtrlQubits)) {
+            if (qubit.getDefiningOp() != parentOp || qubit != parentOutNonCtrlQubits[idx]) {
+                return false;
+            }
+        }
+
+        for (const auto &[idx, qubit] : llvm::enumerate(inCtrlQubits)) {
+            if (qubit.getDefiningOp() != parentOp || qubit != parentOutCtrlQubits[idx]) {
+                return false;
+            }
+        }
+
+        ValueRange opCtrlValues = op.getCtrlValueOperands();
+        ValueRange parentCtrlValues = parentOp.getCtrlValueOperands();
+        if (opCtrlValues.size() != parentCtrlValues.size()) {
+            return false;
+        }
+
+        for (const auto &[idx, v] : llvm::enumerate(opCtrlValues)) {
+            // We assume CSE is already run before this analysis.
+            if (v != parentCtrlValues[idx]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+};
+
+template <typename OpType>
```

Comment:
I think here it is clearly for a custom op

Reply:
Good point, it doesn't need to be templated. I'll change it

Reply:
Actually, it would be good for this class to be able to take in all kinds of ops (custom, unitary, multirz), since otherwise a user of the analysis would do 

```
if (isa<customop>){
    customOpAnalysis;
} else {
    basicVerifyAnalysis;
}
```

Reply:
I suggest keep the template and call it `VerifyParentGateAndNameAnalysis`

Reply:
But in the patterns you already know what you matched.

Reply:
Btw: name sounds good

Reply:
> But in the patterns you already know what you matched.

No, the pattern is a templated type so that all of `customOp`, `unitaryOp` and `multirzOp` can be matched

Reply:
But you use:

```
        if (!isa<quantum::CustomOp>(gate)) {
            // No extra checks for non quantum.custom ops
            return;
        }
```
so it makes no sens to use it for anything else than custom op

Reply:
I see, then I suppose we move this check to the user of the analysis

Reply:
done